cpp/src/arrow/array/data.cc

```diff
@@ -140,6 +141,49 @@ std::shared_ptr<ArrayData> ArrayData::Make(std::shared_ptr<DataType> type, int64
   return std::make_shared<ArrayData>(std::move(type), length, null_count, offset);
 }
 
+namespace {
+template <typename Fn>
+Result<std::shared_ptr<ArrayData>> copy_to_impl(const ArrayData& data,
+                                                const std::shared_ptr<MemoryManager>& to,
+                                                Fn&& copy_fn) {
+  auto output = ArrayData::Make(data.type, data.length, data.null_count, data.offset);
+  output->buffers.reserve(data.buffers.size());
+
+  for (const auto& buf : data.buffers) {
+    if (!buf) {
+      output->buffers.push_back(nullptr);
+      continue;
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto temp_buf, copy_fn(buf, to));
+    output->buffers.push_back(std::move(temp_buf));
+  }
```

Comment:
This is fine, but this can probably be shortened to something like:
```suggestion
  output->buffers.resize(data.buffers.size());
  for (auto& [buf, out_buf] : Zip(data.buffers, output->buffers) {
    if (buf) {
      ARROW_ASSIGN_OR_RAISE(out_buf, copy_fn(buf, to));
    }
  }
```
(you need `arrow/util/range.h` for the `Zip` construct)